packages/client/lib/rpc/modules/eth.ts

```diff
@@ -39,24 +73,178 @@ export class Eth {
+        }
+      }
+    }
+
+    if (!this._vm) {
+      throw new Error('missing vm')
+    }
+
+    if (!transaction.gas) {
+      // If no gas limit is specified use the last block gas limit as an upper bound.
+      const latestHeader = await this._chain.getLatestHeader()
+      transaction.gas = latestHeader.gasLimit as any
+    }
+
+    const vm = this._vm.copy()
+    const txData = { ...transaction, gasLimit: transaction.gas }
+    const tx = Transaction.fromTxData(txData, { common: vm._common, freeze: false })
+
+    // set from address
+    const from = transaction.from ? Address.fromString(transaction.from) : Address.zero()
+    tx.getSenderAddress = () => {
+      return from
+    }
+
+    const { execResult } = await vm.runTx({ tx })
+    return bufferToHex(unpadBuffer(execResult.returnValue))
+  }
+
+  /**
+   * Generates and returns an estimate of how much gas is necessary to allow the transaction to complete.
+   * The transaction will not be added to the blockchain.
+   * Note that the estimate may be significantly more than the amount of gas actually used by the transaction,
+   * for a variety of reasons including EVM mechanics and node performance.
+   * Currently only "latest" block is supported.
+   * @param params An array of two parameters:
+   *   1. The transaction object
+   *       * from (optional) - The address the transaction is sent from
+   *       * to - The address the transaction is directed to
+   *       * gas (optional) - Integer of the gas provided for the transaction execution
+   *       * gasPrice (optional) - Integer of the gasPrice used for each paid gas
+   *       * value (optional) - Integer of the value sent with this transaction
+   *       * data (optional) - Hash of the method signature and encoded parameters.
+   *   2. integer block number, or the string "latest", "earliest" or "pending"
+   * @returns The amount of gas used.
+   */
+  async estimateGas(params: [RpcCallTx, string]) {
+    const [transaction, blockOpt] = params
+
+    if (blockOpt !== 'latest') {
+      const latest = await this.blockNumber()
+      if (blockOpt !== latest) {
+        return {
+          code: INVALID_PARAMS,
+          message: `Currently only "latest" block supported`,
+        }
+      }
+    }
+
+    if (!this._vm) {
+      throw new Error('missing vm')
+    }
+
+    if (!transaction.gas) {
+      // If no gas limit is specified use the last block gas limit as an upper bound.
+      const latestHeader = await this._chain.getLatestHeader()
+      transaction.gas = latestHeader.gasLimit as any
+    }
+
+    const vm = this._vm.copy()
+    const txData = { ...transaction, gasLimit: transaction.gas }
+    const tx = Transaction.fromTxData(txData, { common: vm._common, freeze: false })
+
+    // set from address
+    const from = transaction.from ? Address.fromString(transaction.from) : Address.zero()
+    tx.getSenderAddress = () => {
+      return from
+    }
+
+    const { gasUsed } = await vm.runTx({
```

Comment:
Have to look this up, but `gasUsed` is the total execution gas right, so we did not subtract the gas refund yet?

Reply:
it looks like it is the final amount, docs say `gasUsed = totalGasConsumed - gasRefund`